### DISCUSSION TITLE
managing tw/node instalation without package manager, apt ect

### DISCUSSION THREAD

**Discussion by dubiouscript** [TiddlyWiki/TiddlyWiki5#6963] [Ideas]
id guess that tw on linux users are in the minority 
however hopefully this might be of use to those who like distro hopping or nomadic installs 

why? ....  having a partition for tiddly wiki and related dependency's that's paths and upgrades managed by user independent of distro ect ( at least in theory ;-)

...and  
>  both n itself and all Node.js versions it manages are hosted inside a single, optionally configurable directory,

 the idea being to mount  /tw , on any distro you happen to currently be booting , add the relevant dir's to current path and get a working setup ! 

im working through this so below is the progress so far ,  ( so just n for now , tw eventually when if i figure out how to tweek the default paths ... and read up a bit more )

the instructions as detailed in the repo read me 
https://github.com/Jermolene/TiddlyWiki5#installing-tiddlywiki-on-nodejs

>Install [Node.js](https://tiddlywiki.com/static/Node.js.html)
>
>    Linux:
>
>       Debian/Ubuntu:
>        apt install nodejs
>        May need to be followed up by:
>        apt install npm
>        Arch Linux
>        pacman -S tiddlywiki
>        (installs node and tiddlywiki)

https://github.com/tj/n - Node version management 
 
from the n readme 

>On Linux and macOS, [n-install](https://github.com/mklement0/n-install) allows installation directly from GitHub; for instance:
>
>curl -L https://bit.ly/n-install | bash
>
>n-install sets both PREFIX and N_PREFIX to $HOME/n, installs n to $HOME/n/bin, modifies the initialization files of supported shells to export N_PREFIX and add $HOME/n/bin to the PATH, and installs the latest LTS Node.js version.
>
>As a result, both n itself and all Node.js versions it manages are hosted inside a single, optionally configurable directory, which you >can later remove with the included n-uninstall script. n-update updates n itself to the latest version. See the [n-install repo](https://github.com/mklement0/n-install) for more details.

alternatively , just download the n.sh script into a directory that is all-ready in your environment's $PATH 
and make it executable 

https://raw.githubusercontent.com/tj/n/master/bin/n

then simply use   
> n ls-remote           Output  versions available for download

 or eg to install 
> n lts                      Install the latest LTS Node.js release (downloading if necessary)

.... i hope my skewed  perspective and rambling provides some use/inspiration to some of *nix users *out-there*

**Comment by pmario** [comment]
Do you want to "share" tiddlywiki wikis between the *nix distros or do you also want to share node + TW app. ... 

The later option will probably cause some problems with unix user management. So docker should be an option here. Sharing tw data should be straight forward if you have the data in its own partition. ...

**Reply by dubiouscript** [reply]
hi pmario thanks for your reply ...
i want to do more "distro hopping or nomadic installs"

so the aim is share enviroment ( binary's , paths , directorys )

>"The later option will probably cause some problems with unix user management."

im not shore what problems might be caused , are you thinking of file permitions ??

>" So docker should be an option here." 

tbh : my preff would be : 'guix shell --container' 
ref's 
--> https://10years.guix.gnu.org/video/guix-as-a-tool-for-computational-science
--> https://10years.guix.gnu.org/static/slides/04-hinsen.pdf

some thaughts on why docker should *NOT* be considered 
https://github.com/WeAreWizards/blog/blob/master/content/articles/reproducible-research-or-why-docker-is-not-the-answer.md

>"Sharing tw data should be straight forward if you have the data in its own partition. ..."

im aiming to share the enviroment and decouple it from a running OS & packamanager 
like https://portableapps.com/ but for *nix 

aka thinking out side the box ;)

**Reply by pmario** [reply]
GUIX was new to me. Did watch the video. So it seems to be an environment management system based on docker. Where the working environment is also reproducible. 

So there shouldn't be a "user permission" problem if done right. IMO GUIX may be a solution to work with different unix flavours. If it is available for the distro it may be used that way. 

I think this could work. BUT there may be difficulties using Windows or MacOS, which the majority of TW users use. So you probably miss a big portion (~80%) of potential users.

IMO -- If it doesn't work with windows it doesn't exist.

**Reply by dubiouscript** [reply]
>  "GUIX was new to me. Did watch the video. So it seems to be an environment management system based on docker. "

it is afaik !! a packager manager based on/ inspired by nix (nixos.org) ( nix as written about https://edolstra.github.io/pubs/nixos-icfp2008-submitted.pdf ) ... both nix and guix can run on top of other distros ( they also have there own distros)

>based on docker. 

not exactly (dam the corp branding)  
docker is based on [Linux Namespaces and Cgroups] guix containers afaik use same concepts made available by kernel api(s) 

>IMO -- If it doesn't work with windows it doesn't exist.
:crying_cat_face: https://www.adminspotting.org/  :zipper_mouth_face: :stuck_out_tongue_winking_eye: 

... this is all one step beyond my original goal , which was just dump some random binary's and set the correct dirs in *PATH (manually)  

*https://en.wikipedia.org/wiki/PATH_(variable)

this is (should be) the sort of thing both the nix/guix pkg managers automate 
and the containerization / isolation of environment 
is/will be the icing on the cake  of reproducibility

**Reply by dubiouscript** [reply]
continuing the (somewhatTO) Nocker ( no docker ) tangent 
 https://devclass.com/2022/10/20/docker-quietly-raises-prices-limits-team-accounts
and so begins the digital tupperware lockin 

https://news.ycombinator.com/item?id=33218094
https://github.com/p8952/bocker - Docker implemented in around 100 lines of bash 

 more background 
https://www.theregister.com/2016/07/05/containers_state_of_the_art/

**Reply by dubiouscript** [reply]
> GUIX was new to me. Did watch the video. So it seems to be an environment management system based on docker. Where the working environment is also reproducible.
> 
....
> I think this could work. BUT there may be difficulties using Windows or MacOS, which the majority of TW users use. So you probably miss a big portion (~80%) of potential users.


> IMO -- If it doesn't work with windows it doesn't exist.

sounds like a bad case of "The FUD" 

*groans ghoulishly* 
https://en.wikipedia.org/wiki/Halloween_documents
:ghost: :jack_o_lantern: :scroll: :computer: 


----
how to get Guix running on WSL2 
https://github.com/giuliano108/guix-packages/blob/master/notes/Guix-on-WSL2.md
:stuck_out_tongue_winking_eye: 
----

**Comment by pmario** [comment]
Love that ;)

>>IMO -- If it doesn't work with windows it doesn't exist.
>>😿 https://www.adminspotting.org/ 🤐 😜
>```
>Choose no life. Choose sysadminning. Choose no career.        *****
>Choose no family. Choose a fucking big computer, choose hard  *   *
>disks the size of washing machines, old cars, CD ROM writers  * A *
>and electrical coffee makers. Choose no sleep, high caffeine  * D *
>and mental insurance. Choose fixed interest car loans. Choose * M *
>a rented shoebox. Choose no friends. Choose black jeans and   * I *
>matching combat boots. Choose a swivel chair for your office  * N *
>in a range of fucking fabrics. Choose NNTP and wondering why  * S *
>the fuck you're logged on on a Sunday morning. Choose sitting * P *
>in that chair looking at mind-numbing, spirit-crushing web    * O *
>sites, stuffing fucking junk food into your mouth. Choose     * T *
>rotting away at the end of it all, pishing your last on some  * T *
>miserable newsgroup, nothing more than an embarrassment to    * I *
>the selfish, fucked up lusers Gates spawned to replace the    * N *
>computer-literate.                                            * G *
>Choose your future.                                           *   *
>Choose sysadmining[1].                                        *****
>```

**Reply by dubiouscript** [reply]
it was fairly astute at the time that it was writen *looks back nostagicly* ( to the good old days of 98/me/xp)
**NOW** `we` have more (sm)anthropomorphized(art) phones than people , with difrent corp's hording admin/root access
modern memory hoging browsers 
and all this in wanderfull 4k :upside_down_face: 

and more bandwidth sooo much more band width we need for the 10000 of 3rd party advertising conections (pi hole anyone??)
:zany_face:  :earth_americas: :D
its a funny old world as the say

**Comment by sukima** [comment]
I use TW and Linux every day! One of the things I love about TW is how nomadic it is. I don't have to worry about what environment I'm in since the result is just HTML. For my more complex Wikis where I track via versions control all I need is node which is available for almost every distro has a package for. Or if I need an offline version move the node binary and `node_modules` to a separate drive. Then again if I was going to do that I'd more likely save the TW output HTML file to the drive and update it manually via downloading an `empty.html` from tiddlywiki.com when they release.

All in all the process is about the easiest nomadic setup and with almost zero dependencies. This is a very different situation then most programs which need a package management or a tree of dependencies.

**Reply by dubiouscript** [reply]
> All in all the process is about the easiest nomadic setup and with almost zero dependencies. 

i guess the implicit dependencies are the environment cfg 

idk how you get a node_modules dir with out some pkg manager tbh :thinking: 
... at least i don't see one in the repo :wink: 
>  if I need an offline version move the node binary and `node_modules` to a separate drive.

**Reply by dubiouscript** [reply]
ftr it appears 
examples of tw runing sans node_modules 
can be found hear : https://talk.tiddlywiki.org/t/launching-tiddlywiki-as-a-node-js-app-dependency/6051
:upside_down_face: 
so (theoretically;-) 
with a node binary in $PATH (or whatever equivalent in other OS's  )
and the tiddly wiki repo (via zip download or git clone )  
it is possible to  create  an environment to run TW code 
WITHOUT  apt , npm ect :package: :smiling_face_with_tear: 

weather any one else 
- groks the gist of this topic or 
- cares to try 
 
is another question :rofl: 
as id guess this topic reads like an exert from  
Trev and Simon - "World of the Strange"
:worried: :zany_face: 
---edit---
https://www.youtube.com/watch?v=vT9CR28rJvE :rofl:

**Reply by Jermolene** [reply]
Hi @dubiouscript the intention is indeed for TiddlyWiki to be usable without npm. I think that's important because although npm is taken for granted by developers it is another point of friction for non-developers trying to use TiddlyWiki. I'd welcome any docs improvements to help explain this better.
